contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,443 @@
+volume plugin name (e.g. “google.com/gce-pd”). Note that the CSI driver name can
+be different and this should be just matter of configuration.
+
+Caveat: On downgrade the CSI Driver must still exist to continue the lifecycle
+of the PV’s and volumes it provisioned, the in-tree driver will not pick up
+where the CSI Driver left off. We could have the in-tree volume pick up where
+the CSI driver left off, but in that case the two provisioners race with each
+other if the CSI Driver is still on - this is because there is no consistent way
+(works when CSI driver is both on or off) of communicating between the CSI and
+Internal driver who is in charge of the PV if both contend for ownership of the
+PV.
+
+The CSI Driver replacing the in-tree plugin must support the same storage class
+parameters as the internal volume plugins, so that we continue to have user
+facing consistency. The details of this will not be covered in this design, but
+it will a requirement for admission for migration.
+
+The provisioner will pick up the PVC, and create a PV with a CSI Source that it
+would do normally. Using CSISource in the PV might not be fully “backwards
+compatible” but it will be ok from a user facing behavior change perspective
+because this is a generated PV anyway, we’re not changing user supplied objects.
+The PV is in the domain of the storage administrator and the user should not be
+aware/dependent on what kind of PV ends up getting bound to their PVC (this is a
+core tenet of the k8s storage system).
+
+If the volume was provisioned by the in-tree provisioner but CSI Migration is
+now turned on - On Delete, the internal volume plugin will translate in-tree
+volume source into CSI Source and put it as an annotation to the PV if
+necessary. The external provisioner will check for
+“pv.kubernetes.io/provisioned-by” annotation as usual (added during provisioning
+either by internal or external provisioner) and presence of JSON with CSI
+Source. It will delete the PV only when the PV itself is CSI PV or it has JSON
+annotation with CSI source.
+
+
+### Non-Dynamic Provisioned Volumes (and volumes provisioned before migration)
+In the OperationGenerator at the start of each volume operation call we will
+check to see whether the plugin has been migrated. If it is migrated, we will
+call the in-tree plugin “translation” interface method to get the CSI Source
+JSON.
+
+For Controller calls, we will put the CSI Source JSON into an annotation on the
+PV PV.Annotations[‘k8s.io/CSIVolumeSource‘] and the VolumeToAttach object. This
+requires giving A/D controller and kubelet volume manager PV Update/Patch
+permissions. This is necessary because the CSI external controller components
+all GET the PV object and use it as a source of truth. We also foresee that
+future controller calls such as resize and snapshot will also use and update the
+PV as a source of truth. All these external components will also need to be
```

Comment:
When resizing for example - when user resizes a PVC and volume was originally provisioned by in-tree provisioner then I would assume external controller will update both PV spec and PV CSI spec within the annotation? It is somewhat problematic to update objects as a dependency chain because we don't have transactions..

Reply:
The controller would ensure the annotation is updated/exists before calling CSI resize. The resize function in OperationGenerator would need to be modified to do this as well as call out to CSI instead of in-tree if the "migration" is on.